lib/Interpreter/IncrementalParser.cpp

```diff
@@ -280,6 +280,55 @@ namespace cling {
     return m_Consumer->getTransaction();
   }
 
+  const Transaction*
+  IncrementalParser::mergeTransactionsAfter(const Transaction *T, bool Prev) {
+    llvm::SmallVector<Transaction*, 8> merge;
+    for (std::deque<Transaction*>::reverse_iterator itr = m_Transactions.rbegin(),
+            end  = m_Transactions.rend(); itr != end; ++itr) {
+      Transaction *cur = *itr;
+      merge.push_back(cur);
+      if (cur==T) {
+        if (!Prev)
+          merge.pop_back();
+        break;
+      }
+    }
+    if (merge.size()>1) {
+      llvm::SmallVector<Transaction*, 8>::reverse_iterator itr = merge.rbegin();
+      Transaction *parent = *itr;
+      ++itr;
+      for (llvm::SmallVector<Transaction*, 8>::reverse_iterator end = merge.rend();
+           itr != end; ++itr) {
+        Transaction *cur = *itr;
+        // Try to keep everything current
+        if (m_Consumer->getTransaction()==cur)
+          m_Consumer->setTransaction(parent);
+        if (cur==parent->getNext()) // Make sure parent->next isn't a child
```

Comment:
Because as it loops it is stealing next from whatever it is absorbing.
```
<cling::Transaction* A isEmpty=0 isCommitted=1> 
<cling::Transaction* B isEmpty=0 isCommitted=1> 
<cling::Transaction* C isEmpty=0 isCommitted=1> 
<cling::Transaction* D isEmpty=0 isCommitted=1> 
<cling::Transaction* E isEmpty=0 isCommitted=1> 

A.Next = B, B.Next = C, C.Next = D, D.Next = NULL

Interp->mergeTransactionsAfter(A):

Make B a child: A.Next = B.Next, B.Next = 0
Make C a child: A.Next = C.Next, C.Next = 0
Make D a child: A.Next = D.Next, D.Next = 0
Make D a child: A.Next = E.Next, E.Next = 0
```

If an exception is thrown before all children could be nested, then it should hopefully be in  a state that is valid.